File: engine/encoding/src/main/java/jp/cobolinsight/encoding/CodePage.java
package jp.cobolinsight.encoding;

import java.nio.charset.Charset;
import java.util.Locale;

/** 本ツールが扱うコードページ。 */
public enum CodePage {
    UTF_8("UTF-8", false),
    SHIFT_JIS("windows-31j", false),
    IBM930("x-IBM930", true),
    IBM939("x-IBM939", true);

    private final String charsetName;
    private final boolean ebcdic;

    CodePage(String charsetName, boolean ebcdic) {
        this.charsetName = charsetName;
        this.ebcdic = ebcdic;
    }

    /**
     * コードページ名(別名を含む)を解決する。英数字以外を除いた大小無視の照合で、UTF-8・Shift_JIS・
     * IBM930・IBM939 の別名を受け付ける。未対応の名前は {@link IllegalArgumentException}。
     */
    public static CodePage fromName(String charsetName) {
        String key = charsetName.replaceAll("[^A-Za-z0-9]", "").toLowerCase(Locale.ROOT);
        return switch (key) {
            case "utf8" -> UTF_8;
            case "shiftjis", "sjis", "windows31j", "ms932", "cp932", "932" -> SHIFT_JIS;
            case "ibm930", "xibm930", "cp930", "930" -> IBM930;
            case "ibm939", "xibm939", "cp939", "939" -> IBM939;
            default -> throw new IllegalArgumentException("未対応のコードページ指定: " + charsetName);
        };
    }

    public String charsetName() {
        return charsetName;
    }

    public Charset charset() {
        return Charset.forName(charsetName);
    }

    public boolean isEbcdic() {
        return ebcdic;
    }
}

File: src/gui/src/shared/codepage.ts
/**
 * The codepage vocabulary the GUI offers and the aliases it accepts.
 *
 * Decoding itself belongs to the engine (`decode`), which is the only side that can read EBCDIC
 * CP930/CP939; Node's TextDecoder cannot. This module only normalises the names that travel between
 * the settings screen, the engine arguments and the values recorded in SOURCE.codepage.
 */

/** A codepage the GUI can offer for manual selection. */
export interface CodepageChoice {
  /** The value handed to the engine (--codepage). */
  readonly value: string;
  /** How the name is written in the interface. */
  readonly label: string;
}

/** Selectable codepages, in the order the settings screen lists them. */
export const CODEPAGES: readonly CodepageChoice[] = [
  { value: "Shift_JIS", label: "Shift_JIS" },
  { value: "UTF-8", label: "UTF-8" },
  { value: "IBM930", label: "EBCDIC CP930" },
  { value: "IBM939", label: "EBCDIC CP939" },
];

/**
 * Maps a recorded or hand-entered name onto one of the known codepages, ignoring case and the usual
 * separator and alias differences. Unknown names return null: the caller then shows the raw value
 * rather than guessing an encoding.
 */
export function normalizeCodepage(codepage: string | null | undefined): string | null {
  if (codepage === null || codepage === undefined) {
    return null;
  }
  const key = codepage.trim().toUpperCase().replace(/[_\s-]/g, "");
  switch (key) {
    case "SHIFTJIS":
    case "SJIS":
    case "MS932":
    case "CP932":
    case "WINDOWS31J":
      return "Shift_JIS";
    case "UTF8":
      return "UTF-8";
    case "IBM930":
    case "CP930":
    case "X-IBM930":
    case "XIBM930":
      return "IBM930";
    case "IBM939":
    case "CP939":
    case "XIBM939":
      return "IBM939";
    default:
      return null;
  }
}

/** The display name of a recorded codepage. Unknown values are shown as they were recorded. */
export function codepageLabel(codepage: string | null | undefined, unknownLabel: string): string {
  const normalized = normalizeCodepage(codepage);
  if (normalized === null) {
    return codepage === null || codepage === undefined || codepage === "" ? unknownLabel : codepage;
  }
  return CODEPAGES.find((choice) => choice.value === normalized)?.label ?? normalized;
}

/** Whether the codepage is one of the EBCDIC pair, which carry shift-out/shift-in control bytes. */
export function isEbcdic(codepage: string | null | undefined): boolean {
  const normalized = normalizeCodepage(codepage);
  return normalized === "IBM930" || normalized === "IBM939";
}
